### PR TITLE
Remove LangVer check on using a native int operator

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/64059

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs [hunk 1]
```diff
@@ -11546,6 +11519,27 @@ .maxstack  4
...
+    public static void M()
+    {
+        IntPtr a = default;
+        bool b = a == IntPtr.Zero;
+        a += 1;
+        _ = a + 1;
+    }
+}
+""";
+            var comp = CreateCompilation(source, runtimeFeature: RuntimeFlag.ByRefFields, parseOptions: TestOptions.Regular7_3);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs:11539]: >runtimeFeature: RuntimeFlag.ByRefFields [](http://example.com/codeflow?start=49&length=39)

Why is this needed?

File: src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs [hunk 2]
```diff
@@ -11546,6 +11519,27 @@ .maxstack  4
 ");
         }
 
+        [Fact, WorkItem(63860, "https://github.com/dotnet/roslyn/issues/63860")]
+        public void IntPtrOperatorOnNewPlatform()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs:11523]: >NewPlatform [](http://example.com/codeflow?start=36&length=11)

"NewPlatform" is not a specific term because its meaning is going to change as time goes by.

File: src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs:11539]: >CreateCompilation [](http://example.com/codeflow?start=23&length=17)

What ensures the "NewPlatform" usage?

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 3]
```diff
@@ -41,6 +41,7 @@ public enum RuntimeFlag
         {
             None,
             ByRefFields,
+            NumericIntPtr,
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:44]: >NumericIntPtr [](http://example.com/codeflow?start=12&length=13)

It doesn't look like this enum value is used anywhere, and I don't think we should be adding it. There are plenty of tests that demonstrate the behavior change. If we really have to add a new test, we could use approach that existing tests use, or explicitly target framework that has necessary support (if that is different). I assume this member was meant to be used in ```IntPtrOperatorOnNewPlatform``` unit-test. Somehow the test is passing without it. Is it even testing what it is supposed to?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 4]
```diff
@@ -2730,7 +2727,6 @@ private BoundExpression BindUnaryOperatorCore(CSharpSyntaxNode node, string oper
             UnaryOperatorKind resultOperatorKind = signature.Kind;
             var resultConstant = FoldUnaryOperator(node, resultOperatorKind, resultOperand, resultType, diagnostics);
 
-            CheckNativeIntegerFeatureAvailability(resultOperatorKind, node, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2733]: >CheckNativeIntegerFeatureAvailability [](http://example.com/codeflow?start=12&length=37)

Should the checks be removed conditionally, i.e. only runtime flavor is used?

File: src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs]: Good catch. That was the wrong runtime feature flag. We need to set the NumericIntPtr feature flag to simulate .NET 7 TFM, so that the test would fail absent the code change.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs]: Updated the test name. "new platform" indicates a platform where the numeric IntPtr runtime flag is present.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2733]: Tagging @jaredpar for thought. My understanding is that we want to remove the check unconditionally.
The new compiler comes with a new BCL, so the runtime feature flag is present anyways.

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:44]: Fixed test `IntPtrOperatorOnPlatformWithNumericIntPtrWithLangVer73` to properly recreate the scenario from bug report (new compiler and TFM, but older LangVer) using this flag.
If you prefer, I'd be okay to remove that test and not add overhead to the `CreateCompilation` test helper just for this one scenario. Let me know.

Comment (AlekseyTs) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:44]: I prefer that we do not add a helper just for one scenario.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2733]: > The new compiler comes with a new BCL, so the runtime feature flag is present anyways.

I am not sure if the conclusion is accurate. It is still supported to target older runtimes, I think. In that case, behavior for the native integer feature will be "broken", as changes in NativeIntegerTests,cs indicate.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2733]: It should be fairly easy to check if ```RuntimeSupportsNumericIntPtr``` is true. If so, then the check can be skipped.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

File: src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs [hunk 5]
```diff
@@ -1679,33 +1679,27 @@ .maxstack  2
...
+        IntPtr a = default;
+        bool b = a == IntPtr.Zero;
+        a += 1;
+        _ = a + 1;
+    }
+}
+""";
+            var comp = CreateNumericIntPtrCompilation(source, references: new[] { MscorlibRefWithoutSharingCachedSymbols }, parseOptions: TestOptions.Regular7_3);
+            comp.VerifyDiagnostics();
         }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Emit/NumericIntPtrTests.cs:1703]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing this case:
- with -langversion:7.3 without NumericIntPtr
- with -langversion:latest without NumericIntPtr
- with -langversion:latest with NumericIntPtr

**Review by cston — APPROVED** [review]
